File: kernel/core/mem/slab_allocator.zig

//! SigmaOS Slab Allocator
//! A simple slab memory allocator for kernel use.

const std = @import("std");

/// Slab allocator configuration.
pub const SlabConfig = struct {
    object_size: usize,
    slab_size: usize = 4096, // 4KB per slab
};

/// A single slab in the allocator.
const Slab = struct {
    next: ?*Slab = null,
    free_list: ?*anyopaque = null,
    in_use: usize = 0,
};

/// The slab allocator itself.
pub const SlabAllocator = struct {
    config: SlabConfig,
    slabs: ?*Slab = null,
    empty_slabs: ?*Slab = null,
    allocator: std.mem.Allocator,

    /// Creates a new SlabAllocator.
    pub fn init(allocator: std.mem.Allocator, config: SlabConfig) SlabAllocator {
        return SlabAllocator{
            .config = config,
            .allocator = allocator,
        };
    }

    /// Allocates a new object from the slab allocator.
    pub fn alloc(self: *SlabAllocator) !*anyopaque {
        // Try to find a slab with free space
        var current_slab = self.slabs;
        while (current_slab) |slab| {
            if (slab.free_list != null) {
                const obj = slab.free_list;
                slab.free_list = @ptrCast(*anyopaque, @as([*]u8, @ptrCast(obj))[0..self.config.object_size].ptr);
                slab.in_use += 1;
                return obj;
            }
            current_slab = slab.next;
        }

        // No slab found, allocate a new one
        const new_slab = try self.allocator.create(Slab);
        new_slab.* = .{
            .next = self.slabs,
            .free_list = try self.allocator.allocSentinel(u8, self.config.slab_size, undefined),
            .in_use = 0,
        };

        // Initialize free list
        var free_ptr = @as([*]u8, @ptrCast(new_slab.free_list));
        var i: usize = 0;
        while (i < self.config.slab_size - self.config.object_size) {
            const next_ptr = free_ptr + self.config.object_size;
            @as(*?*anyopaque, @ptrCast(free_ptr)).* = @ptrCast(*anyopaque, next_ptr);
            free_ptr = next_ptr;
            i += self.config.object_size;
        }
        @as(*?*anyopaque, @ptrCast(free_ptr)).* = null;

        self.slabs = new_slab;
        return self.alloc(); // Try allocation again with new slab
    }

    /// Deallocates an object back to the slab allocator.
    pub fn dealloc(self: *SlabAllocator, ptr: *anyopaque) void {
        // Find the slab this object belongs to
        var current_slab = self.slabs;
        var prev_slab: ?*Slab = null;
        while (current_slab) |slab| {
            const slab_start = @intFromPtr(slab.free_list);
            const slab_end = slab_start + self.config.slab_size;
            const obj_addr = @intFromPtr(ptr);

            if (obj_addr >= slab_start and obj_addr < slab_end) {
                // Add back to free list
                @as(*?*anyopaque, @ptrCast(ptr)).* = slab.free_list;
                slab.free_list = ptr;
                slab.in_use -= 1;

                // If slab is empty, move to empty_slabs list
                if (slab.in_use == 0) {
                    if (prev_slab) |prev| {
                        prev.next = slab.next;
                    } else {
                        self.slabs = slab.next;
                    }
                    slab.next = self.empty_slabs;
                    self.empty_slabs = slab;
                }
                return;
            }

            prev_slab = current_slab;
            current_slab = slab.next;
        }
    }
};

File: kernel/core/mem/slab_allocator.rs
//! SigmaOS Slab Allocator
//! Optimized for fixed-size small allocations, implementing the Allocator trait

#![no_std]

use core::ptr::null_mut;
use super::Allocator;

/// Slab Config
pub struct SlabConfig {
    pub object_size: usize,
    pub objects_per_slab: usize,
}

/// A single slab that holds multiple objects
#[repr(C)]
struct Slab {
    next: *mut Slab,
    free_list: *mut u8,
    in_use: usize,
}

/// Slab Allocator - optimized for fixed-size allocations
pub struct SlabAllocator {
    config: SlabConfig,
    slabs: *mut Slab,
    empty_slabs: *mut Slab,
    // For no_std (we'll handle memory manually
}

impl SlabAllocator {
    /// Create a new slab allocator
    pub const fn new(object_size: usize, objects_per_slab: usize) -> Self {
        SlabAllocator {
            config: SlabConfig {
                object_size,
                objects_per_slab,
            },
            slabs: null_mut(),
            empty_slabs: null_mut(),
        }
        // For no_std
    }

    /// Allocator for this allocator, you need to pass a Slab and use a
}

// Implement the Allocator trait (OOP: Polymorphism)
impl Allocator for SlabAllocator {
    fn alloc(&mut self, size: usize) -> *mut u8 {
        // If size doesn't match our object_size, return null (we only handle fixed size!
        if size != self.config.object_size {
            return null_mut();
        }

        // Find a slab with free space
        let mut curr_slab = self.slabs;
        while !curr_slab.is_null() {
            unsafe {
                if !(*curr_slab).free_list.is_null() {
                    let obj = (*curr_slab).free_list;
                    (*curr_slab).free_list = *(obj as *const *mut u8).read(); // Get next free
                    (*curr_slab).in_use += 1;
                    return obj;
                }
            }
            curr_slab = (*curr_slab).next;
        }

        null_mut() // No free slabs or no memory!
    }

    fn dealloc(&mut self, ptr: *mut u8, size: usize) {
        if size != self.config.object_size || ptr.is_null() {
            return;
        }

        // Find the slab for this ptr
        let mut curr_slab = self.slabs;
        let mut prev_slab = null_mut::<*mut Slab>();
        while !curr_slab.is_null() {
            unsafe {
                // Check if ptr is in this slab's memory range
                // We need to know the slab's start and end for this
                // For now, add ptr to the free_list
                // This is a simple implementation
                // Add ptr to the free_list
                *(ptr as *mut *mut u8) = (*curr_slab).free_list;
                (*curr_slab).free_list = ptr;
                (*curr_slab).in_use -= 1;
                break;
            }
            prev_slab = &mut curr_slab;
            curr_slab = (*curr_slab).next;
        }
    }
}

// Safe because we manage memory internally
unsafe impl Send for SlabAllocator {}
unsafe impl Sync for SlabAllocator {}
